### DISCUSSION TITLE
Best Practices for Integrating FastAPI with Separate Library in Python Repository Structure

### DISCUSSION THREAD

**Discussion by scuervo91** [fastapi/fastapi#11454] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
--same repo apprach
my_repository/
│
├── my_library/
│   ├── __init__.py
│   ├── some_module.py
│   └── ...
│   ├── requirements.txt  # Dependencies for the library
│
├── api/
│   ├── __init__.py
│   ├── main.py  # FastAPI code
│   └── ...
│   ├── requirements.txt  # Dependencies for the API
│
└── requirements.txt  # Dependencies for the entire project (optional)

-- different repo approach
fastapi_project/
│
├── __init__.py
├── api/
│   ├── __init__.py
│   ├── main.py  # FastAPI code
│   └── ...
│
└── requirements.txt  # Dependencies for the FastAPI project


my_library/
│
├── __init__.py
├── some_module.py
└── ...
```

### Description

"Seeking Advice: Integrating FastAPI with Independent Library in Python Repository Structure While Maintaining Modularity and Dependencies

Greetings everyone! I'm currently structuring my Python repository for a FastAPI project and need some guidance on maintaining clean code organization and managing dependencies effectively.

Here's the scenario: I've structured my repository with two main components. The first is an independent library (my_library) containing core logic and functionality. This library is designed to be versatile and usable in various contexts, including Jupyter Notebooks. The second component is the FastAPI application (api), which depends on functions and classes from the my_library folder for its logic.

Now, I'm faced with a decision point: should I keep these components within the same repository, or would it be better to split them into two separate repositories? My goal is to ensure that the my_library remains independent with its own dependencies intact, while the FastAPI application (api) can easily access and utilize its functionalities.

If maintaining them within the same repository is the preferred approach, how can I ensure that the FastAPI application (api) depends on the my_library while still keeping their dependencies separate? Would it be advisable to publish my_library to PyPI and have the FastAPI application (api) reference it as a dependency, just like any other external library?

I'm eager to hear your thoughts and experiences on the best practices for managing such a setup. Thank you for your insights!"

### Operating System

macOS

### FastAPI Version

0.101.0

### Pydantic Version

2.7

### Python Version

3.12

**Comment by sehraramiz** [comment]
have you considered to include my_library in the fastapi project as a git submodule?

-- git submodule approach
fastapi_project/
│
├── __init__.py
├── api/
│   ├── __init__.py
│   ├── main.py  # FastAPI code
│   └── ...
├── my_library -> points to my_library original repo
│
└── requirements.txt  # Dependencies for the FastAPI project
└── .gitmodules

my_library/
│
├── __init__.py
├── some_module.py
└── ...

for start you can use the submodule approach and if submodule workflow is working for you then move on to making my_library an independent package on pypi or on your internal organization's infrastructure (if that is the case).
in my experience when you first include a lib as a submodule it enforces you to think more modular and enhance the lib for integration in other projects and develop the lib like a plugin and make it more ready for using it like a python external package 

i don't recommend directly including source code of my_library in any of your other projects repos because it's a headache to maintain changes to my_library across all your projects.

**Comment by jgould22** [comment]
Don't use git submoduals, they just end up being a massive pain in the ass if you do not have a bunch of tooling around them.

> should I keep these components within the same repository, or would it be better to split them into two separate repositories?

Is your library used by more than the fast API? If it is it should be in its own repo.

> while still keeping their dependencies separate? 

What do you mean by this exactly. The dependencies will have to be resolved at some point and they will have to sync up. 

People generally use pyproject.toml and a tool like poetry for this. 

> Would it be advisable to publish my_library to PyPI

Is `my_library` intended to be public code?. If not you would have to host your own index with something like devpi.  If it is publishing to pypi and putting it in its repo is the way to go.

**Comment by YuriiMotov** [comment]
Looks like the author is not interested in this anymore. 
Let's close this for now to keep the discussions clean and organized.
If you’ve found a solution, please share it here.
If you still need help, feel free to let us know.
